algebra/module.lean

```diff
@@ -54,10 +55,308 @@ lemma smul_smul : a • (b • u) = (a * b) • u := mul_smul.symm
+variables {p : set M} [submodule R M p]
+variables {c d : R} (u v w : {x : M // x ∈ p})
+
+include R
+
+protected def add' : {x : M // x ∈ p} → {x : M // x ∈ p} → {x : M // x ∈ p}
+| ⟨v₁, p₁⟩ ⟨v₂, p₂⟩ := ⟨v₁ + v₂, add R p₁ p₂⟩
+
+protected def neg' : {x : M // x ∈ p} → {x : M // x ∈ p}
+| ⟨v, p⟩ := ⟨-v, neg R p⟩
+
+protected def smul' : R → {x : M // x ∈ p} → {x : M // x ∈ p}
+| c ⟨v, p⟩ := ⟨c • v, smul c p⟩
+
+instance : has_add {x : M // x ∈ p} := ⟨submodule.add' R⟩
+instance : has_zero {x : M // x ∈ p} := ⟨⟨0, zero R p⟩⟩
+instance : has_neg {x : M // x ∈ p} := ⟨submodule.neg' R⟩
+instance : has_scalar R {x : M // x ∈ p} := ⟨submodule.smul' R⟩
+
+@[simp] lemma add_val : (u + v).val = u.val + v.val := by cases u; cases v; refl
+@[simp] lemma zero_val : (0 : {x : M // x ∈ p}).val = 0 := rfl
+@[simp] lemma neg_val : (-v).val = -v.val := by cases v; refl
+@[simp] lemma smul_val : (c • v).val = c • v.val := by cases v; refl
+
+lemma sub {u v : M} (HU : u ∈ p) (HV : v ∈ p) : u - v ∈ p := add R HU (neg R HV)
+
+variables (R M p)
+
+instance : module R {x : M // x ∈ p} :=
+{ add                := (+),
+  add_assoc          := λ u v w, subtype.eq (by simp [add_assoc]),
+  zero               := 0,
+  zero_add           := λ v, subtype.eq (by simp [zero_add]),
+  add_zero           := λ v, subtype.eq (by simp [add_zero]),
+  neg                := has_neg.neg,
+  add_left_neg       := λ v, subtype.eq (by simp [add_left_neg]),
+  add_comm           := λ u v, subtype.eq (by simp [add_comm]),
+  smul               := (•),
+  smul_left_distrib  := λ c u v, subtype.eq (by simp [smul_left_distrib]),
+  smul_right_distrib := λ c u v, subtype.eq (by simp [smul_right_distrib]),
+  mul_smul           := λ c d v, subtype.eq (by simp [mul_smul]),
+  one_smul           := λ v, subtype.eq (by simp [one_smul]) }
+
+instance : has_coe (submodule R M p) (module R {x : M // x ∈ p}) := ⟨λ s, submodule.module R M p⟩
+
+end submodule
+
+
+structure linear_map (R M N : Type) [ring R] [module R M] [module R N] :=
+(T : M → N)
+(map_add : ∀ u v, T (u+v) = T u + T v)
+(map_smul : ∀ (c:R) v, T (c • v) = c • (T v))
+
+namespace linear_map
+
+variables {R M N : Type}
+
+section basic
+
+variables [ring R] [module R M] [module R N]
+variables {c d : R} {A B C : linear_map R M N} {u v : M}
+
+instance : has_coe_to_fun (linear_map R M N) := ⟨_, T⟩
+
+@[simp] lemma map_add_app : A (u+v) = A u + A v := A.map_add u v
+@[simp] lemma map_smul_app : A (c•v) = c • (A v) := A.map_smul c v
+
+theorem ext (HAB : ∀ v, A v = B v) : A = B :=
+by {cases A, cases B, congr, exact funext HAB}
+
+@[simp] lemma map_zero : A 0 = 0 :=
+begin
+  have := eq.symm (map_add A 0 0),
+  rw [add_zero] at this,
+  exact eq_zero_of_add_self_eq this
+end
+
+@[simp] lemma map_neg : A (-v) = -(A v) :=
+eq_neg_of_add_eq_zero (by {rw [←map_add_app], simp})
+
+@[simp] lemma map_sub : A (u-v) = A u - A v := by simp
+
+end basic
+
+
+@[simp] def ker [ring R] [module R M] [module R N]
```

Comment:
Why is this marked `simp`? I think that will unfold the definition whenever you call simp, even if that's not the focus of the simplification.

Reply:
Actually it won't unfold the definition even if I mark `simp`, so I'll take it away...